### ISSUE TITLE
VPN connections cause this app to use the remote server instead of the locally available server

### ISSUE THREAD

**Issue by nahaff** [openhab/openhab-android#1111]
### Actual behaviour
When I'm at home, I can connect to VPNs and still get to local servers because only internet traffic goes through the VPN. While on the VPN, I can still reach openHAB through the local IP address in a web browser. However, the openHAB Android app still says "Connecting to remote URL" every time I launch it, and it takes a bit longer. Just to be clear, I'm not trying to connect to my openHAB instance through the VPN. The app is probably just looking at my local IP address information on the VPN interface instead of looking at all interfaces or some other method of testing the local route.

Normally, I rarely use VPN while at home (just for work sometimes), but Google Fi offers always on VPN now. So, it's started to bug me now that it's always on, I'm always "remote".

### Expected behaviour
Open up the app, and it quickly connects to the local URL. Expected even while connected to some VPN, as long is the local instance is really locally available.

 
### Steps to reproduce
1. Connect to VPN. I've used this with OpenVPN and Google's VPN, but I believe results will be similar on any VPN (maybe not a VPN that runs on the local subnet that you're already on, but that's usually only done so VPN can stay always-on and local).
2. Make sure your instance is still locally available. Open OpenHAB in a web browser using the local URL. Most VPNs still allow local LAN traffic.
3. Launch the OpenHAB application and see that it connects to the remote URL instead of the local URL,

I'm not going to fill the rest of this form out, but let me know if you need any more information. I believe it's probably a pretty simple problem, but I'm not sure if the fix is simple. Thanks!

**Comment by mueller-ma** [comment]
This shouldn't happen. If you're using VPN, the app tries to connect to the local URL: https://github.com/openhab/openhab-android/blob/master/mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java#L56-L58

**Comment by kleintody** [comment]
I can confirm a similar problem as mentioned by @nahaff.
I am in my local WLAN. When I use the IP-based URLs (http[s]://my-openhab-ip/), the app says: "Verbinde mit lokaler URL" (i.e. "Connecting with local URL").
However, when I use the DNS-name based URLs (http[s]://my-openhab-hostname/ or http[s]://my-openhab-hostname.my-domainname/), the app says: "Verbinde mit Remote-URL" (i.e.: "Connecting with remote URL").

This is strange, because for my tests I was using always exactly the same URL for the local and remote URL field. Hence, I would expect the app allways to either connect to the local URL or to fail connecting to both - instead, in the latter described situation, the app is connecting to the remote URL but not to the local URL...

**Comment by nahaff** [comment]
Thanks @kleintody !! That will suffice for a work-around in my case, which I'm really glad to have. I'm using HTTPS with a properly signed cert, so I didn't even to think of trying with just the IP address instead of the DNS. It works as expected when the IP address is used, but not with the DNS name. Of course, I had to accept the now invalid cert (because the certs CN/SANs only match DNS names). That'd make me think it's just something with the way my phone handles DNS while on VPN. However, the web browser works fine locally with the DNS name using a web browser, so I think it still counts as a bug. Thanks again!

**Comment by kleintody** [comment]
I think, it’s nothing to do with VPN (or only in parts). As I wrote: this problem occurs for me when I am in my local WLAN of my home. There is no VPN running.

**Comment by kleintody** [comment]
Did you close the ticket?
I think, the IP instead of DNS name is only a workaround. So this still remains a bug for my opinion...

**Comment by mueller-ma** [comment]
If you're using a dns name that cannot be resolved by public dns server, this isn't an openhab bug, as the dns resolution is handled by android itself and not by the app.
However you can buy a domain and create a public dns record that points to your internal openhab server ip.

**Comment by nahaff** [comment]
Yep, I think mueller is correct. I'm guessing you're using something like the cloudlock DNS app? I cannot re-create the issue when I use my own local DNS (without VPN), which is the only DNS that has my openHAB DNS name registered. Ideally for me, Android would check the local DNS first to see if a DNS name resolves to an internal IP address, but that adds latency and security concerns. Registering on public DNS is a good idea for this situation. Should work fine for this app. I'm just now noticing though that my other servers are going through the internet, even when I'm at home. I just didn't notice problems with those yet because they have public DNS names, but that's definitely slowing them down. I can see this being more of a problem for people at work places with a lot of internal servers.

**Comment by kleintody** [comment]
Sorry, but I don‘t get it...
Why should this be an android bug?
1.) I’m directly connected with my home WLAN, not via VPN or so. My local DNS-Server is resolving the domain name properly. Why should android be asking an external nameserver, if the internal (that is deployed via DHCP) is answering DNS requests?
2.) Why is this *identical* DNS-based URL working as an remote but not for local URL? I.E., how can it be that the openHAB app is first failing when trying a „local URL“ XYZ, but in the next step it is successfull when trying exactly the same url XYZ as „remote URL“?
3.) I had never any issues connecting my openHAB server with its DNS name. None of my browsers nor any other app is having issues with using the DNS name of that host. Only the openHAB app has an issue - and, as mentioned, only with using it as local URL, but not as a remote URL.

**Comment by kleintody** [comment]
@nahaff: Which cloudlock DNS?
My openhab instance is running only locally. No connection to any external cloud. 
And why should I register an external DNS if I do not intend to access the host from external, but only internally?

**Comment by nahaff** [comment]
Sounds like we may have different bugs. I'm not saying yours isn't a bug, just that I can't recreate it. Mine seems to work with the DNS name if I just turn off the VPN. Your comments helped me figure it out anyway. The issue I'm having is just from my phone using the VPN's DNS instead of the local one. I can reopen it for you, and would have, but that VPN cause part is right in the title of this bug report, so I don't think it would fit.

**Comment by kleintody** [comment]
P.S.: since the DNS name is only a local one, the app can only get the IP resolved via my local DNS server. So no external source whatsoever should be able to tell the openhab app the internal IP of my internal hostname...
Hence, my question is: how can it be that the app is telling me that it connects to the „remote URL“ but not to the „local URL“, although both are identical and the local should be tried first, shouldn’t it?

**Comment by nahaff** [comment]
@kleintody It's just some app that's been in the Android news lately for a secure DNS option. If you don't know about it, I'd guess you're not using it. OpenHAB should be using the same DNS as a web browser, so if the same URLs aren't working, something something's up. I made the mistake of testing in a web browser with the IP address, but in OpenHAB with the domain name. They operated the same in both for me after I realized that.
